### DISCUSSION TITLE
hb-subset discards mark/mkmk feature?

### DISCUSSION THREAD

**Discussion by madig** [harfbuzz/harfbuzz#2841] [Q&A] [answered]
I want to subset a font while preserving anchors. How do I get hb-subset to stop discarding them?

**Comment by behdad** [comment]
It should work. Unless those lookup types were not implemented (yet). Check PRs. I'll take a look next week.

**Comment by garretrieger** [comment]
First thing, hb-subset by default drops the layout tables so if you're not already doing so add --drop-tables-=GSUB,GPOS,GDEF to your command line.

Mark to Mark lookups should be implemented currently. However there are a few PR's which fixed various issues with layout subsetting still pending so it's possible those may fix the issue. Do you have an example font and the command line for the subset command that you are running?

**Reply by madig** [reply]
That's what I was using, yes.

I'm subsetting https://github.com/bBoxType/FiraSans/blob/master/Fira_Sans_4_3/Fonts/Fira_Sans_TTF_4301/Normal/Roman/FiraSans-Regular.ttf with `hb-subset --text "abc" FiraSans-Regular.ttf -o out.ttf --drop-tables-=GPOS,GSUB,GDEF`. I expect there to be marks for "a". Tested with git master 99184fb.

**Reply by garretrieger** [reply]
I'll give that a try and see what's going on.

**Reply by qxliu76** [reply]
First, this PR is needed to support subsetting extension lookup: #2705 
And then still no marks for "a", which seems to be a usage problem. Both mark glyph and base glyph need to be specified in "--text" option.

**Reply by madig** [reply]
Thanks for the pointer, but I tried that earlier with a bigger subset and just now with `aá` and it also doesn't work.

**Reply by qxliu76** [reply]
"á" is a base glyph not a mark glyph. I checked that fonttools is also dropping markToBase table.

**Reply by madig** [reply]
Yes, but it's a composite of `a` and `acutecomb`, both of which have anchors, or does the subsetter really want the user to specify all wanted glyphs explicitly? I tried again with `--text "́aá"` but ttxing that crashes 🤨

**Reply by qxliu76** [reply]
yes all wanted glyphs need to be specified explicitly. I tried --text "́aá" and it's working fine. No ttx issue for me.

**Reply by madig** [reply]
Ok, now I'm scratching my head. Attaching a Unicode text file with marks at the end: [gs.txt](https://github.com/harfbuzz/harfbuzz/files/5954646/gs.txt) and fontTools 4.19.1. Am on HarfBuzz master, commit feeff5928b7764934cec376d047a5c808e9412a6. Font as linked above.

```text
(venv) /tmp $ hb-subset --text-file /tmp/gs.txt FiraSans-Regular.ttf -o out.ttf --drop-tables-=GSUB,GPOS,GDEF
(venv) /tmp $ ttx -o - out.ttf | rg "mark"
Dumping "out.ttf" to "-"...
Dumping 'GlyphOrder' table...
Dumping 'head' table...
Dumping 'hhea' table...
Dumping 'maxp' table...
Dumping 'OS/2' table...
Dumping 'hmtx' table...
Dumping 'cmap' table...
Dumping 'fpgm' table...
Dumping 'prep' table...
Dumping 'cvt ' table...
Dumping 'loca' table...
Dumping 'glyf' table...
Dumping 'name' table...
Dumping 'post' table...
Dumping 'gasp' table...
Dumping 'GDEF' table...
ERROR: An exception occurred during the decompilation of the 'GPOS' table
Traceback (most recent call last):
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/ttFont.py", line 372, in __getitem__
    return self.tables[tag]
KeyError: 'GPOS'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/ttFont.py", line 391, in __getitem__
    table.decompile(data, self)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otBase.py", line 41, in decompile
    self.table.decompile(reader, font)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otBase.py", line 640, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 547, in read
    table.decompile(reader, font)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otBase.py", line 636, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 167, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 547, in read
    table.decompile(reader, font)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otBase.py", line 636, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 167, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 547, in read
    table.decompile(reader, font)
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otBase.py", line 619, in decompile
    conv = conv.getConverter(reader.tableTag,
  File "/tmp/venv/lib64/python3.9/site-packages/fontTools/ttLib/tables/otConverters.py", line 596, in getConverter
    tableClass = self.lookupTypes[tableType][lookupType]
KeyError: (34, 'SubTable', 'Lookup', 'LookupList')
Dumping 'GPOS' table...
Dumping 'GSUB' table...
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 256 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 257 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 258 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 259 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 260 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
WARNING: name id 261 missing from name table
```

**Reply by madig** [reply]
Or did you mean to say that https://github.com/harfbuzz/harfbuzz/pull/2705 needs to land before this works but you tried it on a simpler font and it worked?

**Reply by garretrieger** [reply]
Yes you'll need #2705 merged for extension lookup subsetting to work, currently on master they are dropped.

**Reply by madig** [reply]
Ah! Great, now I get it :D tried it on LiberationSans-Regular.ttf v2.1.0 and it works as expected. Thanks!

**Comment by behdad** [comment]
@madig for you, pretty much anything you open is probably a bug in our code. So feel free to use Issues instead of Discussion. We don't bite. :)

**Reply by madig** [reply]
Will do after I know what's going on in the log above :scream_cat:

**Comment by behdad** [comment, marked as answer]
So what others said. If you don't specify the mark characters, they won't be included.

What's the alternative? Include all marks that interact with specified characters? That results in a very arbitrary selection. Though I see how that would be useful for production workflows. Feel free to open an issue to discuss the semantics (and we should first add to fonttools subsetter). There has been similar requests in the past by Adam Twardoch and others...

**Reply by madig** [reply]
Hm. I guess that's a philosophical question. You may want to be able to subset `a` and `aacute` and expect `a + acutecomb == aacute` to work. Maybe something for a `--greedy` option.

**Reply by garretrieger** [reply]
We have an issue discussing whether we should do that or not in harfbuzz here: https://github.com/harfbuzz/harfbuzz/issues/2283

**Reply by madig** [reply]
Another thought is that maybe deep glyph exploration by the subsetter becomes less pressing if you can more easily input glyphs you want to keep. Currently, either `--text` or `--text-file` seem to be a hard requirement, you can't just give `--gids`, or even better, actual glyph names (and files).

**Reply by behdad** [reply]
> Currently, either `--text` or `--text-file` seem to be a hard requirement, you can't just give `--gids`, or even better, actual glyph names (and files).

That sure isn't how we want it to work indeed. Fonttools subsetter does the way you suggest and that was as designed. What if you give it an empty `--text` string and `--gids`?

**Reply by madig** [reply]
I was expecting an "either --text or --gids" behavior or maybe union behavior? Anyway, all I really want is https://github.com/harfbuzz/harfbuzz/issues/2455

**Reply by behdad** [reply]
> I was expecting an "either --text or --gids" behavior or maybe union behavior?

Union behavior is how I designed the fonttools subsetter to work. I like to see the same in hb-subset as well.

That is to say, if there's zero characters in the text, we should not produce any GSUB/GPOS/GDEF tables; that is, layout subsetting is automatically dropped because there's no text expected to ever be shaped using the subset font. Internally we don't distinguish between text-only and glyph-only modes. Those should just work as expected given a unified implementation.
